// rax-server-renderer/lib/attribute.js
"use strict";

exports.__esModule = true;
exports.getPropertyInfo = getPropertyInfo;
exports.shouldRemoveAttribute = shouldRemoveAttribute;
exports.POSITIVE_NUMERIC = exports.NUMERIC = exports.OVERLOADED_BOOLEAN = exports.BOOLEAN = exports.BOOLEANISH_STRING = exports.STRING = void 0;
// A simple string attribute.
// Attributes that aren't in the whitelist are presumed to have this type.
var STRING = 1; // A string attribute that accepts booleans in Rax. In HTML, these are called
// "enumerated" attributes with "true" and "false" as possible values.
// When true, it should be set to a "true" string.
// When false, it should be set to a "false" string.

exports.STRING = STRING;
var BOOLEANISH_STRING = 2; // A real boolean attribute.
// When true, it should be present (set either to an empty string or its name).
// When false, it should be omitted.

exports.BOOLEANISH_STRING = BOOLEANISH_STRING;
var BOOLEAN = 3; // An attribute that can be used as a flag as well as with a value.
// When true, it should be present (set either to an empty string or its name).
// When false, it should be omitted.
// For any other value, should be present with that value.

exports.BOOLEAN = BOOLEAN;
var OVERLOADED_BOOLEAN = 4; // An attribute that must be numeric or parse as a numeric.
// When falsy, it should be removed.

exports.OVERLOADED_BOOLEAN = OVERLOADED_BOOLEAN;
var NUMERIC = 5; // An attribute that must be positive numeric or parse as a positive numeric.
// When falsy, it should be removed.

exports.NUMERIC = NUMERIC;
var POSITIVE_NUMERIC = 6;
exports.POSITIVE_NUMERIC = POSITIVE_NUMERIC;
var properties = {};

function getPropertyInfo(prop) {
  return properties[prop];
}

function shouldRemoveAttribute(prop, value) {
  var propertyInfo = getPropertyInfo(prop);
  var propType = propertyInfo ? propertyInfo.type : null;
  var valueType = typeof value;

  if (value === null || valueType === 'undefined') {
    return true;
  }

  switch (valueType) {
    case 'function':
    case 'symbol':
      return true;
  }

  if (propType !== null) {
    switch (propType) {
      case BOOLEAN:
        return !value;

      case OVERLOADED_BOOLEAN:
        return value === false;

      case NUMERIC:
        return isNaN(value);

      case POSITIVE_NUMERIC:
        return isNaN(value) || value < 1;
    }
  }

  return false;
}

['contentEditable', 'draggable', 'spellCheck', 'value'].forEach(function (name) {
  properties[name] = {
    type: BOOLEANISH_STRING
  };
});
['allowFullScreen', 'async', 'autoFocus', 'autoPlay', 'controls', 'default', 'defer', 'disabled', 'disablePictureInPicture', 'formNoValidate', 'hidden', 'loop', 'noModule', 'noValidate', 'open', 'playsInline', 'readOnly', 'required', 'reversed', 'scoped', 'seamless', 'itemScope', 'checked', 'multiple', 'muted', 'selected'].forEach(function (name) {
  properties[name] = {
    type: BOOLEAN
  };
});
['capture', 'download'].forEach(function (name) {
  properties[name] = {
    type: OVERLOADED_BOOLEAN
  };
});
['cols', 'rows', 'size', 'span'].forEach(function (name) {
  properties[name] = {
    type: POSITIVE_NUMERIC
  };
});
