// @alifd/meet-react-component-one/es/recyclerview/VirtualizedList/constants.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

var _scrollProp, _sizeProp, _positionProp, _marginProp, _oppositeMarginProp;

export var ALIGNMENT = {
  AUTO: 'auto',
  START: 'start',
  CENTER: 'center',
  END: 'end'
};
export var DIRECTION = {
  HORIZONTAL: 'horizontal',
  VERTICAL: 'vertical'
};
export var SCROLL_CHANGE_REASON = {
  OBSERVED: 'observed',
  REQUESTED: 'requested'
};
export var scrollProp = (_scrollProp = {}, _defineProperty(_scrollProp, DIRECTION.VERTICAL, 'scrollTop'), _defineProperty(_scrollProp, DIRECTION.HORIZONTAL, 'scrollLeft'), _scrollProp);
export var sizeProp = (_sizeProp = {}, _defineProperty(_sizeProp, DIRECTION.VERTICAL, 'height'), _defineProperty(_sizeProp, DIRECTION.HORIZONTAL, 'width'), _sizeProp);
export var positionProp = (_positionProp = {}, _defineProperty(_positionProp, DIRECTION.VERTICAL, 'top'), _defineProperty(_positionProp, DIRECTION.HORIZONTAL, 'left'), _positionProp);
export var marginProp = (_marginProp = {}, _defineProperty(_marginProp, DIRECTION.VERTICAL, 'marginTop'), _defineProperty(_marginProp, DIRECTION.HORIZONTAL, 'marginLeft'), _marginProp);
export var oppositeMarginProp = (_oppositeMarginProp = {}, _defineProperty(_oppositeMarginProp, DIRECTION.VERTICAL, 'marginBottom'), _defineProperty(_oppositeMarginProp, DIRECTION.HORIZONTAL, 'marginRight'), _oppositeMarginProp);
