// template-component-demo/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _reactRouterDom = require("react-router-dom");

var queryString = _interopRequireWildcard(require("query-string"));

var _Playground = require("./Playground");

var _Layout = require("./Layout");

var _ReactDoc = require("./ReactDoc");

var _excluded = ["demoData"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var App = function App(_ref) {
  var demoData = _ref.demoData,
      readmeData = _ref.readmeData,
      reactDocs = _ref.reactDocs;
  var demos = readmeData ? [].concat(readmeData, demoData) : demoData;
  var items = (demos || []).sort(function (a, b) {
    return a.order - b.order;
  });
  var hasReactDocs = !!(reactDocs && reactDocs.length);
  return /*#__PURE__*/_react["default"].createElement(_Layout.BuildLayout, {
    demos: items,
    reactDocs: reactDocs
  }, items.map(function (data) {
    var Comp = data.component;
    return /*#__PURE__*/_react["default"].createElement("div", {
      key: data.filename
    }, /*#__PURE__*/_react["default"].createElement("a", {
      name: "container_" + data.filename
    }), data.title && data.filename !== 'readme' && /*#__PURE__*/_react["default"].createElement("h3", null, data.title), /*#__PURE__*/_react["default"].createElement(_Playground.Playground, {
      data: data
    }, typeof Comp === 'function' && /*#__PURE__*/_react["default"].createElement(Comp, null)));
  }), hasReactDocs && /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement("h2", null, "API"), /*#__PURE__*/_react["default"].createElement("a", {
    name: "container_api"
  }), reactDocs.map(function (data) {
    return /*#__PURE__*/_react["default"].createElement(_ReactDoc.ReactDoc, {
      key: data.filename,
      data: data
    });
  })));
};

var Demo = function Demo(_ref2) {
  var location = _ref2.location,
      demoData = _ref2.demoData,
      readmeData = _ref2.readmeData;

  var _queryString$parse = queryString.parse(location && location.search),
      demo = _queryString$parse.demo;

  var demos = demoData.find(function (item) {
    return item[0] && item[0].demoKey === demo;
  });
  return /*#__PURE__*/_react["default"].createElement(_Layout.Layout, {
    demos: [].concat(readmeData, demoData),
    matchedFilename: demo
  }, demos ? /*#__PURE__*/_react["default"].createElement(App, {
    demoData: demos
  }) : /*#__PURE__*/_react["default"].createElement(_Playground.Playground, {
    data: readmeData
  }));
};

var AppTab = function AppTab(props) {
  var renderComponent = function renderComponent(routerProps) {
    return /*#__PURE__*/_react["default"].createElement(Demo, (0, _extends2["default"])({}, props, routerProps));
  };

  return /*#__PURE__*/_react["default"].createElement(_reactRouterDom.BrowserRouter, null, /*#__PURE__*/_react["default"].createElement(_reactRouterDom.Switch, null, /*#__PURE__*/_react["default"].createElement(_reactRouterDom.Route, {
    path: "/",
    render: renderComponent
  })));
};

var Main = function Main(_ref3) {
  var demoData = _ref3.demoData,
      rest = (0, _objectWithoutPropertiesLoose2["default"])(_ref3, _excluded);
  var demoLength = demoData.length;
  return demoLength > 1 ? /*#__PURE__*/_react["default"].createElement(AppTab, (0, _extends2["default"])({}, rest, {
    demoData: demoData
  })) : /*#__PURE__*/_react["default"].createElement(App, (0, _extends2["default"])({}, rest, {
    demoData: demoData[0]
  }));
};

var _default = Main;
exports["default"] = _default;
