// @alifd/meet-react/es/side-bar/item.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { __rest } from "tslib";
import classNames from 'classnames';
import React, { createElement, forwardRef, Fragment, useContext, isValidElement } from "react";
import { Text } from "@alifd/meet-react-component-one";
import View from '../view';
import Icon from '../icon';
import Badge from '../badge';
import { SideBarContext } from './side-bar';
import { useGuid } from '../utils/hooks';
import { isString } from '../utils';

var SideBarItem = function SideBarItem(props, ref) {
  var _classNames4;

  var _props$prefix = props.prefix,
      prefix = _props$prefix === void 0 ? 'mt-' : _props$prefix,
      className = props.className,
      itemKey = props.itemKey,
      _props$title = props.title,
      title = _props$title === void 0 ? '' : _props$title,
      _props$icon = props.icon,
      icon = _props$icon === void 0 ? '' : _props$icon,
      _props$count = props.count,
      count = _props$count === void 0 ? 0 : _props$count,
      _props$dot = props.dot,
      dot = _props$dot === void 0 ? false : _props$dot,
      children = props.children,
      _props$onClick = props.onClick,
      onClick = _props$onClick === void 0 ? function () {} : _props$onClick,
      _props$renderContent = props.renderContent,
      renderContent = _props$renderContent === void 0 ? null : _props$renderContent,
      others = __rest(props, ["prefix", "className", "itemKey", "title", "icon", "count", "dot", "children", "onClick", "renderContent"]);

  var _useContext = useContext(SideBarContext),
      activeKey = _useContext.activeKey,
      align = _useContext.align,
      change = _useContext.change;

  var active = "".concat(activeKey) === "".concat(itemKey);
  var clsPrefix = "".concat(prefix, "side-bar-item");
  var curItemId = useGuid('side-bar-item-');
  var clses = {
    bar: classNames(clsPrefix, ["".concat(clsPrefix, "-").concat(align)], _defineProperty({}, "".concat(clsPrefix, "--active"), active), className),
    activeLine: "".concat(clsPrefix, "-active-line"),
    barText: classNames("".concat(clsPrefix, "-text"), _defineProperty({}, "".concat(clsPrefix, "-text--active"), active)),
    barBadge: "".concat(clsPrefix, "-badge"),
    barIcon: classNames("".concat(clsPrefix, "-icon"), _defineProperty({}, "".concat(clsPrefix, "-icon--active"), active))
  };

  var renderTitle = function renderTitle() {
    if (isValidElement(title)) {
      return /*#__PURE__*/React.createElement(View, {
        className: clses.barText
      }, title);
    } else if (isString(title)) {
      return /*#__PURE__*/React.createElement(Text, {
        className: clses.barText
      }, title);
    }

    return null;
  };

  var handleClick = function handleClick(e) {
    if (onClick) {
      onClick(Object.assign(Object.assign({}, e), {
        itemKey: itemKey
      }));
    }

    change(itemKey);
  };

  return /*#__PURE__*/React.createElement(View, _extends({}, others, {
    id: curItemId,
    ref: ref,
    className: clses.bar,
    onClick: handleClick
  }), renderContent ? renderContent({
    title: title,
    active: active,
    icon: icon,
    count: count,
    dot: dot
  }) : dot || count ? /*#__PURE__*/React.createElement(Badge, {
    className: clses.barBadge,
    dot: dot,
    count: count
  }, icon ? /*#__PURE__*/React.createElement(Icon, {
    className: clses.barIcon,
    name: icon
  }) : null, renderTitle()) : /*#__PURE__*/React.createElement(Fragment, null, icon ? /*#__PURE__*/React.createElement(Icon, {
    className: clses.barIcon,
    name: icon
  }) : null, renderTitle()), /*#__PURE__*/React.createElement(View, {
    className: classNames(clses.activeLine, (_classNames4 = {}, _defineProperty(_classNames4, "".concat(clsPrefix, "-active-line--enter"), active), _defineProperty(_classNames4, "".concat(clsPrefix, "-active-line--exit"), !active), _classNames4))
  }));
};

SideBarItem.displayName = 'SideBar.Item';
export default /*#__PURE__*/forwardRef(SideBarItem);
