// @alifd/meet-react-component-one/lib/waterfall/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _view = _interopRequireDefault(require("../view"));

var _scrollview = _interopRequireDefault(require("../scrollview"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function (o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function () { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function (o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var DEFAULT_END_REACHED_THRESHOLD = 500;
var DEFAULT_SCROLL_CALLBACK_THROTTLE = 50;

var Header = /*#__PURE__*/function (_PureComponent) {
  _inherits(Header, _PureComponent);

  var _super = _createSuper(Header);

  function Header() {
    _classCallCheck(this, Header);

    return _super.apply(this, arguments);
  }

  _createClass(Header, [{
    key: "render",
    value: function render() {
      return /*#__PURE__*/_react["default"].createElement(_view["default"], Object.assign({}, this.props));
    }
  }]);

  return Header;
}(_react.PureComponent);

var WebFall = /*#__PURE__*/function (_PureComponent2) {
  _inherits(WebFall, _PureComponent2);

  var _super2 = _createSuper(WebFall);

  function WebFall() {
    var _this;

    _classCallCheck(this, WebFall);

    _this = _super2.apply(this, arguments);

    _this.calcHeightSum = function (arr) {
      var sum = 0;
      arr && arr.forEach(function (item) {
        sum += item;
      });
      return sum;
    };

    return _this;
  }

  _createClass(WebFall, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          _this$props$renderIte = _this$props.renderItem,
          renderItem = _this$props$renderIte === void 0 ? function () {} : _this$props$renderIte,
          dataSource = _this$props.dataSource,
          _this$props$columnCou = _this$props.columnCount,
          columnCount = _this$props$columnCou === void 0 ? 1 : _this$props$columnCou,
          _this$props$leftGap = _this$props.leftGap,
          leftGap = _this$props$leftGap === void 0 ? 0 : _this$props$leftGap,
          _this$props$rightGap = _this$props.rightGap,
          rightGap = _this$props$rightGap === void 0 ? 0 : _this$props$rightGap;
      var columns = [];
      var moduleHeights = [];

      for (var i = 0; i < columnCount; i++) {
        columns[i] = [];
        moduleHeights[i] = 0;
      }

      dataSource && dataSource.forEach(function (item) {
        var targetColumnIndex = 0;
        var minHeight = moduleHeights[0];

        for (var j = 0; j < columnCount; j++) {
          if (moduleHeights[j] < minHeight) {
            minHeight = moduleHeights[j];
            targetColumnIndex = j;
          }
        }

        moduleHeights[targetColumnIndex] += item.height;
        columns[targetColumnIndex].push(item);
      });
      var wrapStyle = Object.assign({}, styles.waterfallWrap, {
        marginLeft: leftGap,
        marginRight: rightGap
      });
      return /*#__PURE__*/_react["default"].createElement(_view["default"], {
        style: wrapStyle
      }, columns.map(function (column, index) {
        return /*#__PURE__*/_react["default"].createElement(_view["default"], {
          key: 'column' + index,
          style: styles.waterfallColumn
        }, column.map(function (item, j) {
          return renderItem(item, 'c_' + index + j);
        }));
      }));
    }
  }]);

  return WebFall;
}(_react.PureComponent);

var Waterfall = /*#__PURE__*/function (_Component) {
  _inherits(Waterfall, _Component);

  var _super3 = _createSuper(Waterfall);

  function Waterfall(props) {
    var _this2;

    _classCallCheck(this, Waterfall);

    _this2 = _super3.call(this, props);
    _this2.loadmoreretry = 1;

    _this2.resetScroll = function () {
      _this2.scrollview.current && _this2.scrollview.current.resetScroll();
    };

    _this2.scrollTo = function (options) {
      _this2.scrollview.current && _this2.scrollview.current.scrollTo(options);
    };

    _this2.state = {
      loadmoreretry: 0
    };
    _this2.scrollview = /*#__PURE__*/(0, _react.createRef)();
    return _this2;
  }

  _createClass(Waterfall, [{
    key: "render",
    value: function render() {
      var props = this.props;
      var renderHeader = props.renderHeader,
          renderFooter = props.renderFooter,
          _props$columnWidth = props.columnWidth,
          columnWidth = _props$columnWidth === void 0 ? 750 : _props$columnWidth;
      var header = typeof renderHeader == 'function' ? renderHeader() : null;
      var footer = typeof renderFooter == 'function' ? renderFooter() : null;
      header = Array.isArray(header) ? header : [header];
      footer = Array.isArray(footer) ? footer : [footer];
      var cells = header.map(function (child, index) {
        if (child) {
          if (child.type != null && child.type != Header) {
            return /*#__PURE__*/_react["default"].createElement(Header, {
              key: 'waterfall_header_' + index
            }, child);
          } else {
            return /*#__PURE__*/(0, _react.cloneElement)(child, {});
          }
        }
      });
      cells = cells.concat( /*#__PURE__*/_react["default"].createElement(WebFall, Object.assign({
        key: 'waterfall_webfall'
      }, props)));
      cells = cells.concat(footer.map(function (child, index) {
        if (child) {
          if (child.type != Header) {
            return /*#__PURE__*/_react["default"].createElement(Header, {
              key: 'waterfall_header_' + index
            }, child);
          } else {
            return /*#__PURE__*/(0, _react.cloneElement)(child, {});
          }
        }
      }));
      styles.waterfallColumn.width = columnWidth;
      return /*#__PURE__*/_react["default"].createElement(_scrollview["default"], Object.assign({}, props, {
        ref: this.scrollview
      }), cells);
    }
  }]);

  return Waterfall;
}(_react.Component);

var styles = {
  waterfallWrap: {
    overflow: 'hidden',
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  waterfallColumn: {}
};
Waterfall.Header = Header;
var _default = Waterfall;
exports["default"] = _default;
