// dingtalk-swap-cli/lib/utils/proxy/rule.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var buffer_1 = require("buffer");
// 本地代理
// anyproxy --intercept --rule proxy.js
function default_1(_a) {
    var bizType = _a.bizType;
    return {
        summary: 'a rule to modify response',
        beforeSendRequest: function (requestDetail) {
            var path, headers, newOption, bizTypeTrans, pcjs, pccss, mobilejs, mobilecss, setterjs, settercss, configjs, cdnPath, newOption;
            return __generator(this, function (_a) {
                path = requestDetail.url;
                headers = requestDetail.requestOptions;
                // 如果是xxx的js，那么就修改为请求本地
                if (path.match(/.*dev\.dingtalk\.com\/main\.bundle\.js/)) {
                    newOption = Object.assign({}, requestDetail.requestOptions, {
                        hostname: '127.0.0.1',
                        port: 3333,
                        headers: __assign({}, headers),
                        path: '/config/index.js',
                    });
                    return [2 /*return*/, {
                            protocol: 'http',
                            requestOptions: newOption
                        }];
                }
                bizTypeTrans = bizType.replace(/\./g, '\\.');
                pcjs = new RegExp("g\\.alicdn\\.com\\/code\\/npm\\/\\@ali\\/dsuite\\-" + bizTypeTrans + "\\/dev\\/pc\\/index\\.js");
                pccss = new RegExp("g\\.alicdn\\.com\\/code\\/npm\\/\\@ali\\/dsuite\\-" + bizTypeTrans + "\\/dev\\/pc\\/index\\.css");
                mobilejs = new RegExp("g\\.alicdn\\.com\\/code\\/npm\\/\\@ali\\/dsuite\\-" + bizTypeTrans + "\\/dev\\/mobile\\/index\\.js");
                mobilecss = new RegExp("g\\.alicdn\\.com\\/code\\/npm\\/\\@ali\\/dsuite\\-" + bizTypeTrans + "\\/dev\\/mobile\\/index\\.css");
                setterjs = new RegExp("g\\.alicdn\\.com\\/code\\/npm\\/\\@ali\\/dsuite\\-" + bizTypeTrans + "\\/dev\\/setter\\/index\\.js");
                settercss = new RegExp("g\\.alicdn\\.com\\/code\\/npm\\/\\@ali\\/dsuite\\-" + bizTypeTrans + "\\/dev\\/setter\\/index\\.css");
                configjs = new RegExp("g\\.alicdn\\.com\\/code\\/npm\\/\\@ali\\/dsuite\\-" + bizTypeTrans + "\\/dev\\/config\\/index\\.js");
                cdnPath = '';
                if (pcjs.test(path)) {
                    cdnPath = '/pc/index.js';
                }
                else if (pccss.test(path)) {
                    cdnPath = '/pc/index.css';
                }
                else if (mobilejs.test(path)) {
                    cdnPath = '/mobile/index.js';
                }
                else if (mobilecss.test(path)) {
                    cdnPath = '/mobile/index.css';
                }
                else if (setterjs.test(path)) {
                    cdnPath = '/setter/index.js';
                }
                else if (settercss.test(path)) {
                    cdnPath = '/setter/index.css';
                }
                else if (configjs.test(path)) {
                    cdnPath = '/config/index.js';
                }
                if (cdnPath) {
                    newOption = Object.assign({}, requestDetail.requestOptions, {
                        hostname: '127.0.0.1',
                        port: 3333,
                        headers: __assign({}, headers),
                        path: cdnPath,
                    });
                    return [2 /*return*/, {
                            protocol: 'http',
                            requestOptions: newOption
                        }];
                }
                return [2 /*return*/];
            });
        },
        beforeSendResponse: function (requestDetail, responseDetail) {
            var path, reqHeaders, newResponse, buf, nRes, script2_1, buf;
            var _a, _b;
            return __generator(this, function (_c) {
                path = requestDetail.url;
                reqHeaders = requestDetail.requestOptions;
                newResponse = responseDetail.response;
                // h5 调试，因为h5，rocket代理，不走http，导致无法通过 html代理，
                // 也不能代理已存在js，因为有dd_cache
                // 只能在页面中加载错误js，来这里调试，通过灰度key来控制js加载
                if (path.match(/.*dev\.dingtalk\.com\/main\.bundle\.js/)) {
                    buf = buffer_1.Buffer.from("\n          ;(function(){\n            if(typeof eruda == 'undefined'){\n              var parent = document.head || document.body || document.documentElement;\n              var script = document.createElement('script');\n              script.type = 'text/javascript';\n              script.crossOrigin = 'anonymous';\n              script.src = '//g.alicdn.com/code/lib/eruda/2.4.1/eruda.min.js';\n              script.onload = function () {\n                eruda.init();\n              };\n              parent.appendChild(script);\n            }\n          })();\n          window.__SUITE_DEV_MICRO={bizType:'" + bizType + "',version:'dev'};\n          ");
                    newResponse.body = buf;
                    return [2 /*return*/, new Promise(function (resolve, reject) {
                            resolve({ response: newResponse, });
                        })];
                }
                // pc 和 designer
                if ((path.indexOf('aflow.dingtalk.com/dingtalk/pc/query/pchomepage.htm') > -1
                    /*||path.indexOf('aflow.dingtalk.com/dingtalk/mobile/homepage.htm') > -1 */
                    || path.indexOf('aflow.dingtalk.com/dingtalk/web/query/oaDesigner') > -1)
                    && ((_a = newResponse.header['Content-Type']) === null || _a === void 0 ? void 0 : _a.indexOf('text/html')) > -1
                    && ((_b = reqHeaders.headers['Accept']) === null || _b === void 0 ? void 0 : _b.indexOf('html')) > -1) {
                    nRes = newResponse.body.toString();
                    script2_1 = "\n              <script src=\"https://g.alicdn.com/code/lib/eruda/2.4.1/eruda.min.js\"></script>\n              <script>\n                window.__SUITE_DEV_MICRO={bizType:'" + bizType + "',version:'dev'};\n                eruda.init();\n              </script>\n            ";
                    nRes = nRes.replace(/<head>/i, function (i) {
                        return i + script2_1;
                    });
                    buf = buffer_1.Buffer.from(nRes);
                    newResponse.body = buf;
                    return [2 /*return*/, new Promise(function (resolve, reject) {
                            resolve({ response: newResponse, });
                        })];
                }
                return [2 /*return*/];
            });
        },
    };
}
exports.default = default_1;
