// @uni/video/es/ali-miniapp/chooseMedia.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import { CONTAINER_NAME } from "../_utils/constant";
import { normalize } from "../common";
import showActionSheet from "@uni/action-sheet/lib/ali-miniapp/index.js";
var chooseMedia = normalize.chooseMedia(function (args) {
  var mediaType = args.mediaType;

  var imageFn = function imageFn() {
    return my.chooseImage(_extends({}, args, {
      success: function success(res) {
        var result = {
          type: 'image',
          tempFiles: res.tempFiles.map(function (i) {
            return _extends({}, i, {
              tempFilePath: i.path
            });
          })
        };
        args.success(result);
      }
    }));
  };

  var videoFn = function videoFn() {
    return my.chooseVideo(_extends({}, args, {
      success: function success(res) {
        var result = {
          type: 'video',
          tempFiles: [res]
        };
        args.success(result);
      }
    }));
  };

  if (mediaType.length === 1 && mediaType[0] === 'image') {
    imageFn();
  } else if (mediaType.length === 1 && mediaType[0] === 'video') {
    videoFn();
  } else {
    showActionSheet({
      itemList: ['图片', '视频']
    }).then(function (res) {
      if (res.tapIndex === 0) {
        imageFn();
      } else if (res.tapIndex === 1) {
        videoFn();
      } else {
        return Promise.reject({
          errMsg: '用户取消选择'
        });
      }
    });
  }
}, CONTAINER_NAME.ALIPAY);
export default chooseMedia;
