// template-component-demo/lib/Playground/CodeBox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.CodeBox = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireDefault(require("react"));

var _codeboxModule = _interopRequireDefault(require("./codebox.module.scss"));

var CodeBox = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2["default"])(CodeBox, _React$Component);

  function CodeBox(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _this.handleChange = function () {
      _this.setState(function (prevState) {
        return {
          check: !prevState.check
        };
      });
    };

    _this.state = {
      check: false
    };
    return _this;
  }

  var _proto = CodeBox.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        filename = _this$props.filename,
        children = _this$props.children,
        highlightedCode = _this$props.highlightedCode,
        highlightedStyle = _this$props.highlightedStyle;
    var check = this.state.check;
    if (!children) return null;
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: _codeboxModule["default"].codeBox
    }, filename && children && /*#__PURE__*/_react["default"].createElement("div", {
      className: _codeboxModule["default"].demo
    }, /*#__PURE__*/_react["default"].createElement("div", {
      id: filename
    }, children)), /*#__PURE__*/_react["default"].createElement("div", {
      className: _codeboxModule["default"].actions
    }, /*#__PURE__*/_react["default"].createElement("button", {
      type: "button",
      title: "Close editor",
      onClick: this.handleChange
    }, /*#__PURE__*/_react["default"].createElement("svg", {
      xmlns: "http://www.w3.org/2000/svg",
      width: "15",
      height: "24",
      viewBox: "0 0 24 24",
      fill: "none",
      stroke: "currentColor",
      strokeWidth: "2",
      strokeLinecap: "round",
      strokeLinejoin: "round"
    }, /*#__PURE__*/_react["default"].createElement("polyline", {
      points: "16 18 22 12 16 6"
    }), /*#__PURE__*/_react["default"].createElement("polyline", {
      points: "8 6 2 12 8 18"
    })))), /*#__PURE__*/_react["default"].createElement("div", {
      className: "markdown code-wrapper",
      style: {
        display: check ? 'block' : 'none'
      }
    }, highlightedCode && /*#__PURE__*/_react["default"].createElement("div", {
      className: "highlight"
    }, /*#__PURE__*/_react["default"].createElement("pre", null, /*#__PURE__*/_react["default"].createElement("code", {
      lang: "jsx",
      dangerouslySetInnerHTML: {
        __html: highlightedCode
      }
    }))), highlightedStyle && /*#__PURE__*/_react["default"].createElement("div", {
      className: "highlight"
    }, /*#__PURE__*/_react["default"].createElement("pre", null, /*#__PURE__*/_react["default"].createElement("code", {
      lang: "css",
      dangerouslySetInnerHTML: {
        __html: highlightedStyle
      }
    })))));
  };

  return CodeBox;
}(_react["default"].Component);

exports.CodeBox = CodeBox;
