// @uni/clipboard/lib/common.js
"use strict";

exports.__esModule = true;
exports.normalizeGet = normalizeGet;
exports.normalizeSet = normalizeSet;

var _promisify = require("./_utils/promisify");

var _styleOptions = require("./_utils/styleOptions");

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/**
 * actionsheet返回值格式化
 * @param res
 */
var formatResponse = function formatResponse(res) {
  return {
    text: res.data != undefined ? res.data : res.text
  };
};
/**
 * actionsheet参数格式化
 * @param options
 */


function styleOptions(options) {
  return options ? _extends({}, options, {
    success: function success(res) {
      options.success && options.success(formatResponse(res));
    }
  }) : {};
}

function normalizeGet(api, containerName) {
  return function (args) {
    return (0, _promisify.promisify)(api)(styleOptions((0, _styleOptions.styleIn)(args, containerName))).then(formatResponse);
  };
}

function normalizeSet(api, containerName) {
  return function (args) {
    return (0, _promisify.promisify)(api)((0, _styleOptions.styleIn)(args, containerName));
  };
}
