// template-component-demo/lib/Layout/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Layout = exports.BuildLayout = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireDefault(require("react"));

var _Sidebar = _interopRequireDefault(require("./Sidebar"));

var _indexModule = _interopRequireDefault(require("./index.module.scss"));

var Layout = function Layout(_ref) {
  var demos = _ref.demos,
      children = _ref.children,
      matchedFilename = _ref.matchedFilename;
  return /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement("div", {
    className: _indexModule["default"].nav
  }, "Docs"), /*#__PURE__*/_react["default"].createElement("div", {
    className: _indexModule["default"].container
  }, /*#__PURE__*/_react["default"].createElement(_Sidebar["default"], {
    items: demos,
    matchedFilename: matchedFilename
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: _indexModule["default"].main
  }, children)));
};

exports.Layout = Layout;

var BuildLayout = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2["default"])(BuildLayout, _React$Component);

  function BuildLayout(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _this.handleChangeHash = function () {
      setTimeout(function () {
        // update currentHash after location changed
        _this.setState({
          currentHash: window.location.hash
        });
      }, 0);
    };

    _this.state = {
      currentHash: window.location.hash
    };
    return _this;
  }

  var _proto = BuildLayout.prototype;

  _proto.render = function render() {
    var _this2 = this;

    var _this$props = this.props,
        demos = _this$props.demos,
        children = _this$props.children,
        reactDocs = _this$props.reactDocs;
    var currentHash = this.state.currentHash;
    return /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement("div", {
      className: _indexModule["default"].layout
    }, children), /*#__PURE__*/_react["default"].createElement("ul", {
      className: _indexModule["default"].fixNav
    }, demos.map(function (item) {
      return /*#__PURE__*/_react["default"].createElement("li", {
        key: item.filename,
        className: currentHash.replace('#container_', '') === item.filename ? _indexModule["default"].selected : ''
      }, /*#__PURE__*/_react["default"].createElement("a", {
        href: "#container_" + item.filename,
        onClick: _this2.handleChangeHash
      }, item.title || item.filename));
    }), reactDocs && reactDocs.length > 0 && /*#__PURE__*/_react["default"].createElement("li", null, /*#__PURE__*/_react["default"].createElement("a", {
      href: "#container_api",
      onClick: this.handleChangeHash
    }, "API"))));
  };

  return BuildLayout;
}(_react["default"].Component);

exports.BuildLayout = BuildLayout;
