// template-component-demo/lib/Layout/Sidebar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireDefault(require("react"));

var _reactRouterDom = require("react-router-dom");

var _sidebarModule = _interopRequireDefault(require("./sidebar.module.scss"));

var Sidebar = /*#__PURE__*/function (_React$Component) {
  (0, _inheritsLoose2["default"])(Sidebar, _React$Component);

  function Sidebar(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _this.handleChange = function (evt) {
      var value = evt.target.value || '';

      var menus = _this.props.items.filter(function (menu) {
        var item = Array.isArray(menu) ? menu[0] : menu;
        var title = (item.demoKey || item.title).toLocaleLowerCase();
        var val = value.toLocaleLowerCase();
        return title.includes(val);
      });

      _this.setState({
        value: value,
        menus: menus
      });
    };

    _this.state = {
      value: '',
      menus: props.items
    };
    return _this;
  }

  var _proto = Sidebar.prototype;

  _proto.render = function render() {
    var matchedFilename = this.props.matchedFilename;
    var _this$state = this.state,
        value = _this$state.value,
        menus = _this$state.menus;
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: _sidebarModule["default"].sidebar
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: _sidebarModule["default"].searchbox,
      style: {
        display: 'none'
      }
    }, /*#__PURE__*/_react["default"].createElement("input", {
      type: "text",
      className: _sidebarModule["default"].input,
      placeholder: "Type to search...",
      value: value,
      onChange: this.handleChange
    })), /*#__PURE__*/_react["default"].createElement("dl", {
      className: _sidebarModule["default"].menuList
    }, menus.map(function (menu, key) {
      var item = Array.isArray(menu) ? menu[0] : menu;
      return /*#__PURE__*/_react["default"].createElement("dt", {
        className: "" + _sidebarModule["default"].item + (matchedFilename === (item.demoKey || item.filename) ? " " + _sidebarModule["default"].selected : ''),
        key: (item.demoKey || item.filename) + "-" + key
      }, /*#__PURE__*/_react["default"].createElement(_reactRouterDom.Link, {
        to: "?demo=" + (item.demoKey || item.filename)
      }, item.demoKey || item.title));
    })));
  };

  return Sidebar;
}(_react["default"].Component);

exports["default"] = Sidebar;
