// @alifd/meet-react-component-one/lib/picture/pictureDefault.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _view = _interopRequireDefault(require("../view"));

var _image = _interopRequireDefault(require("../image"));

var _universalEnv = require("universal-env");

var _universalDevice = require("universal-device");

var _index = _interopRequireDefault(require("./optimizer/index"));

var _webp = require("./webp");

var _rpx2vw = _interopRequireDefault(require("./rpx2vw"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var isSupportJPG = false;
var isSupportPNG = false;

if (!_universalEnv.isNode) {
  (0, _webp.isSupport)(function (_isSupportJPG) {
    isSupportJPG = _isSupportJPG;
  });
  (0, _webp.isSupport)(function (_isSupportPNG) {
    isSupportPNG = _isSupportPNG;
  }, 'alpha');
}

function parseSuffix(suffix) {
  var result = [];
  var ret = [];

  if (typeof suffix === 'string') {
    ret = suffix.split(',');
  }

  if (Array.isArray(suffix)) {
    ret = suffix;
  }

  if (ret && ret[0]) {
    result[0] = ret[0];
  }

  if (ret && ret[1]) {
    result[1] = ret[1];
  }

  return result;
}

function getQualitySuffix(highQuality, suffix) {
  var _suffix = parseSuffix(suffix);

  return highQuality ? _suffix[0] : _suffix[1];
}

var Picture = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var children = props.children,
      className = props.className,
      _props$style = props.style,
      style = _props$style === void 0 ? {} : _props$style,
      resizeMode = props.resizeMode,
      width = props.width,
      height = props.height,
      _props$placeholder = props.placeholder,
      placeholder = _props$placeholder === void 0 ? 'data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgYGBgAAAABQABh6FO1AAAAABJRU5ErkJggg==' : _props$placeholder,
      _props$source = props.source,
      source = _props$source === void 0 ? {
    uri: ''
  } : _props$source,
      _props$autoRemoveSche = props.autoRemoveScheme,
      autoRemoveScheme = _props$autoRemoveSche === void 0 ? true : _props$autoRemoveSche,
      _props$autoReplaceDom = props.autoReplaceDomain,
      autoReplaceDomain = _props$autoReplaceDom === void 0 ? true : _props$autoReplaceDom,
      _props$autoScaling = props.autoScaling,
      autoScaling = _props$autoScaling === void 0 ? true : _props$autoScaling,
      _props$autoWebp = props.autoWebp,
      autoWebp = _props$autoWebp === void 0 ? true : _props$autoWebp,
      _props$ignoreGif = props.ignoreGif,
      ignoreGif = _props$ignoreGif === void 0 ? true : _props$ignoreGif,
      _props$autoCompress = props.autoCompress,
      autoCompress = _props$autoCompress === void 0 ? true : _props$autoCompress,
      _props$highQuality = props.highQuality,
      highQuality = _props$highQuality === void 0 ? true : _props$highQuality,
      _props$compressSuffix = props.compressSuffix,
      compressSuffix = _props$compressSuffix === void 0 ? ['Q75', 'Q50'] : _props$compressSuffix,
      _props$defaultHeight = props.defaultHeight,
      defaultHeight = _props$defaultHeight === void 0 ? '100vw' : _props$defaultHeight,
      _props$lazyload = props.lazyload,
      lazyload = _props$lazyload === void 0 ? false : _props$lazyload,
      _props$autoPixelRatio = props.autoPixelRatio,
      autoPixelRatio = _props$autoPixelRatio === void 0 ? true : _props$autoPixelRatio;
  var uri = source.uri;
  var nativeProps = Object.assign({}, props);

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      visible = _useState2[0],
      setVisible = _useState2[1];

  var sWidth = style.width,
      sHeight = style.height;

  if (!sHeight && sWidth && width && height) {
    var pScaling = width / (typeof sWidth === 'string' ? parseInt(sWidth, 10) : sWidth);
    sHeight = parseInt(height / pScaling + '', 10);
  }

  var newStyle = Object.assign({
    height: sHeight
  }, style);

  if (uri) {
    if (!_universalEnv.isNode && autoPixelRatio && _universalDevice.devicePixelRatio > 1) {
      if (typeof sWidth === 'string' && sWidth.indexOf('rpx') > -1) {
        sWidth = (0, _rpx2vw["default"])(parseInt(sWidth.split('rpx')[0]) * 2 + 'rpx');
      }
    }

    uri = (0, _index["default"])(uri, {
      ignoreGif: ignoreGif,
      ignorePng: true,
      removeScheme: autoRemoveScheme,
      replaceDomain: autoReplaceDomain,
      scalingWidth: autoScaling ? sWidth : 0,
      webp: autoWebp && isSupportJPG && isSupportPNG,
      compressSuffix: autoCompress ? getQualitySuffix(highQuality, compressSuffix) : ''
    });
  }

  if (resizeMode) {
    newStyle.resizeMode = resizeMode;
  }

  var url = placeholder;

  if (lazyload) {
    nativeProps.onAppear = function () {
      return setVisible(true);
    };

    if (visible) {
      url = uri;
    }
  } else {
    url = uri;
  }

  if (children) {
    return /*#__PURE__*/_react["default"].createElement(_view["default"], Object.assign({}, nativeProps, {
      className: className,
      style: Object.assign(Object.assign({}, newStyle), {
        backgroundImage: 'url(' + url + ')',
        backgroundSize: resizeMode || 'cover',
        backgroundRepeat: 'no-repeat',
        backgroundPosition: resizeMode === 'cover' || resizeMode === 'contain' ? 'center' : null,
        height: newStyle.height ? newStyle.height : defaultHeight
      }),
      "data-once": true
    }), children);
  } else {
    return /*#__PURE__*/_react["default"].createElement(_image["default"], Object.assign({}, nativeProps, {
      ref: ref,
      className: className,
      style: newStyle,
      "data-once": true,
      source: {
        uri: url
      }
    }));
  }
});
Picture.displayName = 'Picture';
var _default = Picture;
exports["default"] = _default;
