// rax-children/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _rax = require("rax");

var flattenChildren = _rax.shared.flattenChildren;

function convertChildrenToArray(children) {
  // flatten children
  children = flattenChildren(children, []);
  return Array.isArray(children) ? children : [].concat(children);
}

var Children = {
  map: function map(children, fn, ctx) {
    if (children == null) return null;
    children = convertChildrenToArray(children);
    return children.map(function (child, index) {
      return fn.call(ctx, child, index);
    });
  },
  forEach: function forEach(children, fn, ctx) {
    if (children == null) return null;
    children = convertChildrenToArray(children);
    children.forEach(function (child, index) {
      return fn.call(ctx, child, index);
    });
  },
  count: function count(children) {
    if (children == null) return 0;
    return convertChildrenToArray(children).length;
  },
  only: function only(children) {
    // `only` receive rax element child
    // null value is not acceptable
    children = Children.toArray(children);
    if (children.length !== 1) throw new Error('Children.only: expected to receive a single element child.');
    return children[0];
  },
  toArray: function toArray(children) {
    if (children == null) return []; // `toArray` filter null value

    return convertChildrenToArray(children).filter(function (child) {
      return child !== null;
    });
  }
};
var _default = Children;
exports.default = _default;
