// @alifd/meet-react/lib/progress/utils.js
"use strict";

exports.__esModule = true;
exports.statusColor = exports.progressiveColor = exports.progressSize = exports.progressCircleWidthSize = exports.progressCircleFontSize = exports.progressBackgroundColor = exports.getValue = exports.addRpx = void 0;

var _utils = require("../utils");

var addRpx = function (value) {
  if (!value) {
    return value;
  }

  return (0, _utils.isString)(value) ? value : value + "rpx";
};

exports.addRpx = addRpx;

var getValue = function (cssProp) {
  if (!cssProp) {
    return 0;
  }

  if ((0, _utils.isNumber)(cssProp)) {
    return cssProp;
  }

  var index = cssProp.indexOf('rpx');

  if (index === -1) {
    return Number(cssProp);
  }

  return Number(cssProp.substring(0, index));
};

exports.getValue = getValue;

var statusColor = function (status) {
  return {
    normal: '#209BFA',
    error: '#EB2F2F',
    success: '#05D161'
  }[status];
};

exports.statusColor = statusColor;

var progressBackgroundColor = function (color) {
  return color || '#f2f2f2';
};

exports.progressBackgroundColor = progressBackgroundColor;

var progressSize = function (size) {
  return {
    small: 4,
    medium: 6,
    large: 8
  }[size] || size;
};

exports.progressSize = progressSize;

var progressiveColor = function (percent) {
  if (percent < 33) {
    return '#EB2F2F';
  }

  if (percent > 66) {
    return '#05D161';
  }

  return '#D97B00';
};

exports.progressiveColor = progressiveColor;

var progressCircleWidthSize = function (size) {
  return {
    small: 22,
    medium: 28,
    large: 34
  }[size] || size;
};

exports.progressCircleWidthSize = progressCircleWidthSize;

var progressCircleFontSize = function (size) {
  return {
    small: 14,
    medium: 16,
    large: 18
  }[size] || size;
};

exports.progressCircleFontSize = progressCircleFontSize;
