// stylesheet-loader/lib/BoxModelPropTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PropTypes_1 = require("./PropTypes");
var BoxModelPropTypes = {
    width: PropTypes_1.default.length,
    height: PropTypes_1.default.length,
    minWidth: PropTypes_1.default.length,
    minHeight: PropTypes_1.default.length,
    maxWidth: PropTypes_1.default.length,
    maxHeight: PropTypes_1.default.length,
    marginLeft: PropTypes_1.default.length,
    marginRight: PropTypes_1.default.length,
    marginTop: PropTypes_1.default.length,
    marginBottom: PropTypes_1.default.length,
    paddingLeft: PropTypes_1.default.length,
    paddingRight: PropTypes_1.default.length,
    paddingTop: PropTypes_1.default.length,
    paddingBottom: PropTypes_1.default.length,
    borderWidth: PropTypes_1.default.length,
    borderLeftWidth: PropTypes_1.default.length,
    borderTopWidth: PropTypes_1.default.length,
    borderRightWidth: PropTypes_1.default.length,
    borderBottomWidth: PropTypes_1.default.length,
    borderLeftStyle: PropTypes_1.default.oneOf(['dotted', 'dashed', 'solid']),
    borderTopStyle: PropTypes_1.default.oneOf(['dotted', 'dashed', 'solid']),
    borderRightStyle: PropTypes_1.default.oneOf(['dotted', 'dashed', 'solid']),
    borderBottomStyle: PropTypes_1.default.oneOf(['dotted', 'dashed', 'solid']),
    borderStyle: PropTypes_1.default.oneOf(['dotted', 'dashed', 'solid']),
    borderRadius: PropTypes_1.default.length,
    borderBottomLeftRadius: PropTypes_1.default.length,
    borderBottomRightRadius: PropTypes_1.default.length,
    borderTopLeftRadius: PropTypes_1.default.length,
    borderTopRightRadius: PropTypes_1.default.length,
    backgroundImage: PropTypes_1.default.string,
    overflow: PropTypes_1.default.oneOf(['hidden', 'visible']),
    position: PropTypes_1.default.oneOf(['relative', 'absolute', 'sticky', 'fixed']),
    display: PropTypes_1.default.oneOf(['block', 'flex', 'inline-flex', 'inline-block', 'inline']),
    top: PropTypes_1.default.length,
    bottom: PropTypes_1.default.length,
    left: PropTypes_1.default.length,
    right: PropTypes_1.default.length,
    opacity: PropTypes_1.default.number,
    transform: PropTypes_1.default.string,
};
exports.default = BoxModelPropTypes;
