// universal-device/lib/index.js
"use strict";

exports.__esModule = true;
exports.rpx2px = exports.px2rpx = exports.screenHeight = exports.screenWidth = exports.platform = exports.appName = void 0;

var _universalEnv = require("universal-env");

var _web = _interopRequireDefault(require("./web"));

var _weex = _interopRequireDefault(require("./weex"));

var _ali = _interopRequireDefault(require("./miniapp/ali"));

var _wechat = _interopRequireDefault(require("./miniapp/wechat"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function dutyChain() {
  for (var _len = arguments.length, fns = new Array(_len), _key = 0; _key < _len; _key++) {
    fns[_key] = arguments[_key];
  }

  for (var i = 0; i < fns.length; i++) {
    var result = fns[i]();

    if (result) {
      return result;
    }
  }
}

function handleWeb() {
  if (_universalEnv.isWeb) {
    return _web.default;
  }

  return null;
}

function handleWeex() {
  if (_universalEnv.isWeex) {
    return _weex.default;
  }

  return null;
}

function handleMiniApp() {
  if (_universalEnv.isMiniApp) {
    return _ali.default;
  }

  return null;
}

function handleWeChat() {
  if (_universalEnv.isWeChatMiniProgram) {
    return _wechat.default;
  }

  return null;
}

var deviceInfo = dutyChain(handleWeb, handleWeex, handleMiniApp, handleWeChat);
var appName = deviceInfo.appName;
exports.appName = appName;
var platform = deviceInfo.platform;
exports.platform = platform;
var screenWidth = deviceInfo.screenWidth;
exports.screenWidth = screenWidth;
var screenHeight = deviceInfo.screenHeight;
exports.screenHeight = screenHeight;

function wrapper(fn, screenWidth) {
  return function (value) {
    return fn(screenWidth, value);
  };
}

var CALCULATION_ACCURACY = 8;
var px2rpx = wrapper(function (screenWidth, value) {
  return Number((750 * value / screenWidth).toFixed(CALCULATION_ACCURACY));
}, screenWidth);
exports.px2rpx = px2rpx;
var rpx2px = wrapper(function (screenWidth, value) {
  return Number((screenWidth / 750 * value).toFixed(CALCULATION_ACCURACY));
}, screenWidth);
exports.rpx2px = rpx2px;
