// @uni/video/lib/ali-miniapp/chooseMedia.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _constant = require("../_utils/constant");

var _common = require("../common");

var _index = _interopRequireDefault(require("@uni/action-sheet/lib/ali-miniapp/index.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var chooseMedia = _common.normalize.chooseMedia(function (args) {
  var mediaType = args.mediaType;

  var imageFn = function imageFn() {
    return my.chooseImage(_extends({}, args, {
      success: function success(res) {
        var result = {
          type: 'image',
          tempFiles: res.tempFiles.map(function (i) {
            return _extends({}, i, {
              tempFilePath: i.path
            });
          })
        };
        args.success(result);
      }
    }));
  };

  var videoFn = function videoFn() {
    return my.chooseVideo(_extends({}, args, {
      success: function success(res) {
        var result = {
          type: 'video',
          tempFiles: [res]
        };
        args.success(result);
      }
    }));
  };

  if (mediaType.length === 1 && mediaType[0] === 'image') {
    imageFn();
  } else if (mediaType.length === 1 && mediaType[0] === 'video') {
    videoFn();
  } else {
    (0, _index.default)({
      itemList: ['图片', '视频']
    }).then(function (res) {
      if (res.tapIndex === 0) {
        imageFn();
      } else if (res.tapIndex === 1) {
        videoFn();
      } else {
        return Promise.reject({
          errMsg: '用户取消选择'
        });
      }
    });
  }
}, _constant.CONTAINER_NAME.ALIPAY);

var _default = chooseMedia;
exports.default = _default;
