// @alifd/meet-react/es/progress/utils.js
import { isNumber, isString } from '../utils';
export var addRpx = function addRpx(value) {
  if (!value) {
    return value;
  }

  return isString(value) ? value : "".concat(value, "rpx");
};
export var getValue = function getValue(cssProp) {
  if (!cssProp) {
    return 0;
  }

  if (isNumber(cssProp)) {
    return cssProp;
  }

  var index = cssProp.indexOf('rpx');

  if (index === -1) {
    return Number(cssProp);
  }

  return Number(cssProp.substring(0, index));
};
export var statusColor = function statusColor(status) {
  var colorMap = {
    normal: '#209BFA',
    error: '#EB2F2F',
    success: '#05D161'
  };
  return colorMap[status];
};
export var progressBackgroundColor = function progressBackgroundColor(color) {
  return color || '#f2f2f2';
};
export var progressSize = function progressSize(size) {
  var sizeMap = {
    small: 4,
    medium: 6,
    large: 8
  };
  return sizeMap[size] || size;
};
export var progressiveColor = function progressiveColor(percent) {
  if (percent < 33) {
    return '#EB2F2F';
  }

  if (percent > 66) {
    return '#05D161';
  }

  return '#D97B00';
};
export var progressCircleWidthSize = function progressCircleWidthSize(size) {
  var sizeMap = {
    small: 22,
    medium: 28,
    large: 34
  };
  return sizeMap[size] || size;
};
export var progressCircleFontSize = function progressCircleFontSize(size) {
  var sizeMap = {
    small: 14,
    medium: 16,
    large: 18
  };
  return sizeMap[size] || size;
};
