// template-component-demo/es/index.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/extends";
var _excluded = ["demoData"];

/* eslint-disable jsx-a11y/anchor-has-content */
import React from 'react';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import * as queryString from 'query-string';
import { Playground } from './Playground';
import { BuildLayout, Layout } from './Layout';
import { ReactDoc } from './ReactDoc';

var App = function App(_ref) {
  var demoData = _ref.demoData,
      readmeData = _ref.readmeData,
      reactDocs = _ref.reactDocs;
  var demos = readmeData ? [].concat(readmeData, demoData) : demoData;
  var items = (demos || []).sort(function (a, b) {
    return a.order - b.order;
  });
  var hasReactDocs = !!(reactDocs && reactDocs.length);
  return /*#__PURE__*/React.createElement(BuildLayout, {
    demos: items,
    reactDocs: reactDocs
  }, items.map(function (data) {
    var Comp = data.component;
    return /*#__PURE__*/React.createElement("div", {
      key: data.filename
    }, /*#__PURE__*/React.createElement("a", {
      name: "container_" + data.filename
    }), data.title && data.filename !== 'readme' && /*#__PURE__*/React.createElement("h3", null, data.title), /*#__PURE__*/React.createElement(Playground, {
      data: data
    }, typeof Comp === 'function' && /*#__PURE__*/React.createElement(Comp, null)));
  }), hasReactDocs && /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("h2", null, "API"), /*#__PURE__*/React.createElement("a", {
    name: "container_api"
  }), reactDocs.map(function (data) {
    return /*#__PURE__*/React.createElement(ReactDoc, {
      key: data.filename,
      data: data
    });
  })));
};

var Demo = function Demo(_ref2) {
  var location = _ref2.location,
      demoData = _ref2.demoData,
      readmeData = _ref2.readmeData;

  var _queryString$parse = queryString.parse(location && location.search),
      demo = _queryString$parse.demo;

  var demos = demoData.find(function (item) {
    return item[0] && item[0].demoKey === demo;
  });
  return /*#__PURE__*/React.createElement(Layout, {
    demos: [].concat(readmeData, demoData),
    matchedFilename: demo
  }, demos ? /*#__PURE__*/React.createElement(App, {
    demoData: demos
  }) : /*#__PURE__*/React.createElement(Playground, {
    data: readmeData
  }));
};

var AppTab = function AppTab(props) {
  var renderComponent = function renderComponent(routerProps) {
    return /*#__PURE__*/React.createElement(Demo, _extends({}, props, routerProps));
  };

  return /*#__PURE__*/React.createElement(Router, null, /*#__PURE__*/React.createElement(Switch, null, /*#__PURE__*/React.createElement(Route, {
    path: "/",
    render: renderComponent
  })));
};

var Main = function Main(_ref3) {
  var demoData = _ref3.demoData,
      rest = _objectWithoutPropertiesLoose(_ref3, _excluded);

  var demoLength = demoData.length;
  return demoLength > 1 ? /*#__PURE__*/React.createElement(AppTab, _extends({}, rest, {
    demoData: demoData
  })) : /*#__PURE__*/React.createElement(App, _extends({}, rest, {
    demoData: demoData[0]
  }));
};

export default Main;
