// @uni/navigate/lib/common.js
"use strict";

exports.__esModule = true;
exports.normalize = void 0;

var _promisify = require("./_utils/promisify");

var _styleOptions = require("./_utils/styleOptions");

var normalize = {
  push: function push(api, containerName) {
    return function (options) {
      return (0, _promisify.promisify)(api)((0, _styleOptions.styleIn)(options, containerName));
    };
  },
  back: function back(api, containerName) {
    return function (options) {
      return (0, _promisify.promisify)(api)((0, _styleOptions.styleIn)(options, containerName));
    };
  },
  go: function go(api, containerName) {
    return function (options) {
      return (0, _promisify.promisify)(api)((0, _styleOptions.styleIn)(options, containerName));
    };
  },
  replace: function replace(api, containerName) {
    return function (options) {
      return (0, _promisify.promisify)(api)((0, _styleOptions.styleIn)(options, containerName));
    };
  },
  reLaunch: function reLaunch(api, containerName) {
    return function (options) {
      return (0, _promisify.promisify)(api)((0, _styleOptions.styleIn)(options, containerName));
    };
  },
  switchTab: function switchTab(api, containerName) {
    return function (options) {
      return (0, _promisify.promisify)(api)((0, _styleOptions.styleIn)(options, containerName));
    };
  }
};
exports.normalize = normalize;
