// @uni/navigate/lib/index.js
"use strict";

exports.__esModule = true;
exports.switchTab = exports.replace = exports.reLaunch = exports.push = exports.go = exports.default = exports.back = void 0;

var _env = require("@uni/env");

var _index = _interopRequireDefault(require("./ali-miniapp/index"));

var _index2 = _interopRequireDefault(require("./wechat-miniprogram/index"));

var _index3 = _interopRequireDefault(require("./bytedance-microapp/index"));

var _index4 = _interopRequireDefault(require("./web/index"));

var _index5 = _interopRequireDefault(require("./kuaishou-miniprogram/index"));

var _index6 = _interopRequireDefault(require("./baidu-smartprogram/index"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var push = function push(options) {
  if (_env.isWeb) {
    return _index4.default.push(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return _index5.default.push(options);
  } else if (_env.isWeChatMiniProgram) {
    return _index2.default.push(options);
  } else if (_env.isByteDanceMicroApp) {
    return _index3.default.push(options);
  } else if (_env.isMiniApp) {
    return _index.default.push(options);
  } else if (_env.isBaiduSmartProgram) {
    return _index6.default.push(options);
  } else {
    throw new Error('Uni API：navigate.push 暂不支持');
  }
};

exports.push = push;

var go = function go(options) {
  if (_env.isWeb) {
    return _index4.default.go(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return _index5.default.go(options);
  } else if (_env.isWeChatMiniProgram) {
    return _index2.default.go(options);
  } else if (_env.isByteDanceMicroApp) {
    return _index3.default.go(options);
  } else if (_env.isMiniApp) {
    return _index.default.go(options);
  } else if (_env.isBaiduSmartProgram) {
    return _index6.default.go(options);
  } else {
    throw new Error('Uni API：navigate.go 暂不支持');
  }
};

exports.go = go;

var back = function back(options) {
  if (_env.isWeb) {
    return _index4.default.back(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return _index5.default.back(options);
  } else if (_env.isWeChatMiniProgram) {
    return _index2.default.back(options);
  } else if (_env.isByteDanceMicroApp) {
    return _index3.default.back(options);
  } else if (_env.isMiniApp) {
    return _index.default.back(options);
  } else if (_env.isBaiduSmartProgram) {
    return _index6.default.back(options);
  } else {
    throw new Error('Uni API：navigate.back 暂不支持');
  }
};

exports.back = back;

var replace = function replace(options) {
  if (_env.isWeb) {
    return _index4.default.replace(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return _index5.default.replace(options);
  } else if (_env.isWeChatMiniProgram) {
    return _index2.default.replace(options);
  } else if (_env.isByteDanceMicroApp) {
    return _index3.default.replace(options);
  } else if (_env.isMiniApp) {
    return _index.default.replace(options);
  } else if (_env.isBaiduSmartProgram) {
    return _index6.default.replace(options);
  } else {
    throw new Error('Uni API：navigate.replace 暂不支持');
  }
};

exports.replace = replace;

var reLaunch = function reLaunch(options) {
  if (_env.isWeb) {
    return _index4.default.reLaunch(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return _index5.default.reLaunch(options);
  } else if (_env.isWeChatMiniProgram) {
    return _index2.default.reLaunch(options);
  } else if (_env.isByteDanceMicroApp) {
    return _index3.default.reLaunch(options);
  } else if (_env.isMiniApp) {
    return _index.default.reLaunch(options);
  } else if (_env.isBaiduSmartProgram) {
    return _index6.default.reLaunch(options);
  } else {
    throw new Error('Uni API：navigate.reLaunch 暂不支持');
  }
};

exports.reLaunch = reLaunch;

var switchTab = function switchTab(options) {
  if (_env.isWeb) {
    return _index4.default.switchTab(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return _index5.default.switchTab(options);
  } else if (_env.isWeChatMiniProgram) {
    return _index2.default.switchTab(options);
  } else if (_env.isByteDanceMicroApp) {
    return _index3.default.switchTab(options);
  } else if (_env.isMiniApp) {
    return _index.default.switchTab(options);
  } else if (_env.isBaiduSmartProgram) {
    return _index6.default.switchTab(options);
  } else {
    throw new Error('Uni API：navigate.switchTab 暂不支持');
  }
};

exports.switchTab = switchTab;
var _default = {
  push: push,
  back: back,
  reLaunch: reLaunch,
  switchTab: switchTab,
  replace: replace,
  go: go
};
exports.default = _default;
