// @uni/file/lib/web/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _openDocument = _interopRequireDefault(require("./openDocument"));

exports.openDocument = _openDocument.default;

var _removeSaved = _interopRequireDefault(require("./removeSaved"));

exports.removeSaved = _removeSaved.default;

var _upload = _interopRequireDefault(require("./upload"));

exports.upload = _upload.default;

var _save = _interopRequireDefault(require("./save"));

exports.save = _save.default;

var _getSavedList = _interopRequireDefault(require("./getSavedList"));

exports.getSavedList = _getSavedList.default;

var _getSavedInfo = _interopRequireDefault(require("./getSavedInfo"));

exports.getSavedInfo = _getSavedInfo.default;

var _getInfo = _interopRequireDefault(require("./getInfo"));

exports.getInfo = _getInfo.default;

var _download = _interopRequireDefault(require("./download"));

exports.download = _download.default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = {
  openDocument: _openDocument.default,
  removeSaved: _removeSaved.default,
  upload: _upload.default,
  save: _save.default,
  getSavedList: _getSavedList.default,
  getSavedInfo: _getSavedInfo.default,
  getInfo: _getInfo.default,
  download: _download.default
};
exports.default = _default;
