// @alifd/meet-react/lib/tab/item.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _tslib = require("tslib");

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireWildcard(require("react"));

var _meetReactComponentOne = require("@alifd/meet-react-component-one");

var _view = _interopRequireDefault(require("../view"));

var _icon = _interopRequireDefault(require("../icon"));

var _tab = require("./tab");

var _hooks = require("../utils/hooks");

var _utils = require("../utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var TabItem = function (props, ref) {
  var _classNames, _classNames2, _classNames3, _classNames4;

  var _props$prefix = props.prefix,
      prefix = _props$prefix === void 0 ? 'mt-' : _props$prefix,
      className = props.className,
      size = props.size,
      itemKey = props.itemKey,
      _props$title = props.title,
      title = _props$title === void 0 ? '' : _props$title,
      _props$icon = props.icon,
      icon = _props$icon === void 0 ? '' : _props$icon,
      children = props.children,
      _props$onClick = props.onClick,
      onClick = _props$onClick === void 0 ? function () {} : _props$onClick,
      _props$renderContent = props.renderContent,
      renderContent = _props$renderContent === void 0 ? null : _props$renderContent,
      others = (0, _tslib.__rest)(props, ["prefix", "className", "size", "itemKey", "title", "icon", "children", "onClick", "renderContent"]);

  var _useContext = (0, _react.useContext)(_tab.TabContext),
      activeKey = _useContext.activeKey,
      change = _useContext.change,
      fullWidth = _useContext.fullWidth,
      shape = _useContext.shape,
      activeLineType = _useContext.activeLineType;

  var active = "" + activeKey === "" + itemKey;
  var clsPrefix = prefix + "tab-item";
  var curItemId = (0, _hooks.useGuid)('tab-item-');
  var clses = {
    tab: (0, _classnames.default)(clsPrefix, clsPrefix + "--" + size, clsPrefix + "--" + shape, (_classNames = {}, _classNames[clsPrefix + "--active"] = active, _classNames[clsPrefix + "--active-" + shape] = active, _classNames[clsPrefix + "--fullwidth"] = fullWidth, _classNames), className),
    activeLine: clsPrefix + "-active-line " + clsPrefix + "-active-line--" + activeLineType + " " + clsPrefix + "-active-line--" + activeLineType + "-" + size,
    tabIcon: (0, _classnames.default)(clsPrefix + "-icon", clsPrefix + "-icon--" + size, clsPrefix + "-icon--" + shape, (_classNames2 = {}, _classNames2[clsPrefix + "-icon--active"] = active, _classNames2[clsPrefix + "-icon--active-" + shape] = active, _classNames2)),
    tabText: (0, _classnames.default)(clsPrefix + "-text", clsPrefix + "-text--" + size, clsPrefix + "-text--" + shape, (_classNames3 = {}, _classNames3[clsPrefix + "-text--active"] = active, _classNames3[clsPrefix + "-text--active-" + size] = active && shape !== 'capsule', _classNames3[clsPrefix + "-text--active-" + shape] = active, _classNames3))
  };
  return /*#__PURE__*/(0, _react.createElement)(_view.default, _extends({}, others, {
    id: curItemId,
    ref: ref,
    className: clses.tab,
    onClick: function handleClick(e) {
      if (onClick) {
        onClick(Object.assign(Object.assign({}, e), {
          itemKey: itemKey
        }));
      }

      change(itemKey);
    }
  }), renderContent ? renderContent({
    title: title,
    icon: icon,
    active: active
  }) : /*#__PURE__*/(0, _react.createElement)(_react.Fragment, null, icon ? /*#__PURE__*/(0, _react.createElement)(_icon.default, {
    className: clses.tabIcon,
    name: icon
  }) : null, function renderTitle() {
    if ((0, _react.isValidElement)(title)) {
      return /*#__PURE__*/(0, _react.createElement)(_view.default, {
        className: clses.tabText
      }, title);
    } else if ((0, _utils.isString)(title)) {
      return /*#__PURE__*/(0, _react.createElement)(_meetReactComponentOne.Text, {
        className: clses.tabText
      }, title);
    }

    return null;
  }()), shape !== 'capsule' && /*#__PURE__*/(0, _react.createElement)(_view.default, {
    className: (0, _classnames.default)(clses.activeLine, (_classNames4 = {}, _classNames4[clsPrefix + "-active-line--enter"] = active, _classNames4[clsPrefix + "-active-line--exit"] = !active, _classNames4))
  }));
};

TabItem.displayName = 'Tab.Item';

var _default = /*#__PURE__*/(0, _react.forwardRef)(TabItem);

exports.default = _default;
