// @uni/file/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = exports.openDocument = exports.removeSaved = exports.save = exports.getSavedList = exports.getSavedInfo = exports.getInfo = exports.download = exports.upload = void 0;

var _env = require("@uni/env");

var aliMiniAppModule = _interopRequireWildcard(require("./ali-miniapp/index"));

var weChatModule = _interopRequireWildcard(require("./wechat-miniprogram/index"));

var bytedanceModule = _interopRequireWildcard(require("./bytedance-microapp/index"));

var webModule = _interopRequireWildcard(require("./web/index"));

var kuaiShouModule = _interopRequireWildcard(require("./kuaishou-miniprogram/index"));

var baiDuModule = _interopRequireWildcard(require("./baidu-smartprogram/index"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var upload = function upload(options) {
  if (_env.isWeb) {
    return webModule.upload(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return kuaiShouModule.upload(options);
  } else if (_env.isWeChatMiniProgram) {
    return weChatModule.upload(options);
  } else if (_env.isByteDanceMicroApp) {
    return bytedanceModule.upload(options);
  } else if (_env.isMiniApp) {
    return aliMiniAppModule.upload(options);
  } else if (_env.isBaiduSmartProgram) {
    return baiDuModule.upload(options);
  } else {
    throw new Error('@uni：upload暂不支持');
  }
};

exports.upload = upload;

var download = function download(options) {
  if (_env.isWeb) {
    return webModule.download(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return kuaiShouModule.download(options);
  } else if (_env.isWeChatMiniProgram) {
    return weChatModule.download(options);
  } else if (_env.isByteDanceMicroApp) {
    return bytedanceModule.download(options);
  } else if (_env.isMiniApp) {
    return aliMiniAppModule.download(options);
  } else if (_env.isBaiduSmartProgram) {
    return baiDuModule.download(options);
  } else {
    throw new Error('@uni：download暂不支持');
  }
};

exports.download = download;

var getInfo = function getInfo(options) {
  if (_env.isWeb) {
    return webModule.getInfo(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return kuaiShouModule.getInfo(options);
  } else if (_env.isWeChatMiniProgram) {
    return weChatModule.getInfo(options);
  } else if (_env.isByteDanceMicroApp) {
    return bytedanceModule.getInfo(options);
  } else if (_env.isMiniApp) {
    return aliMiniAppModule.getInfo(options);
  } else if (_env.isBaiduSmartProgram) {
    return baiDuModule.getInfo(options);
  } else {
    throw new Error('@uni：getInfo暂不支持');
  }
};

exports.getInfo = getInfo;

var getSavedInfo = function getSavedInfo(options) {
  if (_env.isWeb) {
    return webModule.getSavedInfo(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return kuaiShouModule.getSavedInfo(options);
  } else if (_env.isWeChatMiniProgram) {
    return weChatModule.getSavedInfo(options);
  } else if (_env.isByteDanceMicroApp) {
    return bytedanceModule.getSavedInfo(options);
  } else if (_env.isMiniApp) {
    return aliMiniAppModule.getSavedInfo(options);
  } else if (_env.isBaiduSmartProgram) {
    return baiDuModule.getSavedInfo(options);
  } else {
    throw new Error('@uni：getSavedInfo暂不支持');
  }
};

exports.getSavedInfo = getSavedInfo;

var getSavedList = function getSavedList(options) {
  if (_env.isWeb) {
    return webModule.getSavedList(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return kuaiShouModule.getSavedList(options);
  } else if (_env.isWeChatMiniProgram) {
    return weChatModule.getSavedList(options);
  } else if (_env.isByteDanceMicroApp) {
    return bytedanceModule.getSavedList(options);
  } else if (_env.isMiniApp) {
    return aliMiniAppModule.getSavedList(options);
  } else if (_env.isBaiduSmartProgram) {
    return baiDuModule.getSavedList(options);
  } else {
    throw new Error('@uni：getSavedList暂不支持');
  }
};

exports.getSavedList = getSavedList;

var save = function save(options) {
  if (_env.isWeb) {
    return webModule.save(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return kuaiShouModule.save(options);
  } else if (_env.isWeChatMiniProgram) {
    return weChatModule.save(options);
  } else if (_env.isByteDanceMicroApp) {
    return bytedanceModule.save(options);
  } else if (_env.isMiniApp) {
    return aliMiniAppModule.save(options);
  } else if (_env.isBaiduSmartProgram) {
    return baiDuModule.save(options);
  } else {
    throw new Error('@uni：save暂不支持');
  }
};

exports.save = save;

var removeSaved = function removeSaved(options) {
  if (_env.isWeb) {
    return webModule.removeSaved(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return kuaiShouModule.removeSaved(options);
  } else if (_env.isWeChatMiniProgram) {
    return weChatModule.removeSaved(options);
  } else if (_env.isByteDanceMicroApp) {
    return bytedanceModule.removeSaved(options);
  } else if (_env.isMiniApp) {
    return aliMiniAppModule.removeSaved(options);
  } else if (_env.isBaiduSmartProgram) {
    return baiDuModule.removeSaved(options);
  } else {
    throw new Error('@uni：removeSaved暂不支持');
  }
};

exports.removeSaved = removeSaved;

var openDocument = function openDocument(options) {
  if (_env.isWeb) {
    return webModule.openDocument(options);
  } else if (_env.isKuaiShouMiniProgram) {
    return kuaiShouModule.openDocument(options);
  } else if (_env.isWeChatMiniProgram) {
    return weChatModule.openDocument(options);
  } else if (_env.isByteDanceMicroApp) {
    return bytedanceModule.openDocument(options);
  } else if (_env.isMiniApp) {
    return aliMiniAppModule.openDocument(options);
  } else if (_env.isBaiduSmartProgram) {
    return baiDuModule.openDocument(options);
  } else {
    throw new Error('@uni：openDocument暂不支持');
  }
};

exports.openDocument = openDocument;
var _default = {
  openDocument: openDocument,
  removeSaved: removeSaved,
  upload: upload,
  save: save,
  getSavedList: getSavedList,
  getSavedInfo: getSavedInfo,
  getInfo: getInfo,
  download: download
};
exports.default = _default;
