// @uni/image/lib/web/chooseImage.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _constant = require("../_utils/constant");

var _common = require("../common");

var _chooseMedia = _interopRequireDefault(require("@uni/video/lib/web/chooseMedia.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chooseImage = _common.normalize.chooseImage(function (args) {
  if (args === void 0) {
    args = {};
  }

  var _args = args,
      _args$count = _args.count,
      count = _args$count === void 0 ? 1 : _args$count,
      _args$success = _args.success,
      success = _args$success === void 0 ? function () {} : _args$success,
      _args$fail = _args.fail,
      fail = _args$fail === void 0 ? function () {} : _args$fail,
      _args$complete = _args.complete,
      complete = _args$complete === void 0 ? function () {} : _args$complete;
  (0, _chooseMedia.default)({
    count: count,
    accept: 'image/*'
  }).then(function (res) {
    var result = {
      tempFilePaths: res.tempFiles.map(function (item) {
        return item.tempFilePath;
      }),
      tempFiles: res.tempFiles.map(function (item) {
        return {
          path: item.tempFilePath,
          size: item.size
        };
      }),
      files: res.files
    };
    success(result);
    complete(result);
  }).catch(function (e) {
    fail(e);
    complete(e);
  });
}, _constant.CONTAINER_NAME.WEB);

var _default = chooseImage;
exports.default = _default;
