// template-component-demo/es/Playground/CodeBox.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React from 'react';
import styles from './codebox.module.scss';
export var CodeBox = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(CodeBox, _React$Component);

  function CodeBox(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _this.handleChange = function () {
      _this.setState(function (prevState) {
        return {
          check: !prevState.check
        };
      });
    };

    _this.state = {
      check: false
    };
    return _this;
  }

  var _proto = CodeBox.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        filename = _this$props.filename,
        children = _this$props.children,
        highlightedCode = _this$props.highlightedCode,
        highlightedStyle = _this$props.highlightedStyle;
    var check = this.state.check;
    if (!children) return null;
    return /*#__PURE__*/React.createElement("div", {
      className: styles.codeBox
    }, filename && children && /*#__PURE__*/React.createElement("div", {
      className: styles.demo
    }, /*#__PURE__*/React.createElement("div", {
      id: filename
    }, children)), /*#__PURE__*/React.createElement("div", {
      className: styles.actions
    }, /*#__PURE__*/React.createElement("button", {
      type: "button",
      title: "Close editor",
      onClick: this.handleChange
    }, /*#__PURE__*/React.createElement("svg", {
      xmlns: "http://www.w3.org/2000/svg",
      width: "15",
      height: "24",
      viewBox: "0 0 24 24",
      fill: "none",
      stroke: "currentColor",
      strokeWidth: "2",
      strokeLinecap: "round",
      strokeLinejoin: "round"
    }, /*#__PURE__*/React.createElement("polyline", {
      points: "16 18 22 12 16 6"
    }), /*#__PURE__*/React.createElement("polyline", {
      points: "8 6 2 12 8 18"
    })))), /*#__PURE__*/React.createElement("div", {
      className: "markdown code-wrapper",
      style: {
        display: check ? 'block' : 'none'
      }
    }, highlightedCode && /*#__PURE__*/React.createElement("div", {
      className: "highlight"
    }, /*#__PURE__*/React.createElement("pre", null, /*#__PURE__*/React.createElement("code", {
      lang: "jsx",
      dangerouslySetInnerHTML: {
        __html: highlightedCode
      }
    }))), highlightedStyle && /*#__PURE__*/React.createElement("div", {
      className: "highlight"
    }, /*#__PURE__*/React.createElement("pre", null, /*#__PURE__*/React.createElement("code", {
      lang: "css",
      dangerouslySetInnerHTML: {
        __html: highlightedStyle
      }
    })))));
  };

  return CodeBox;
}(React.Component);
