// @alifd/meet-react-component-one/lib/recyclerview/VirtualizedList/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sizeProp = exports.scrollProp = exports.positionProp = exports.oppositeMarginProp = exports.marginProp = exports.SCROLL_CHANGE_REASON = exports.DIRECTION = exports.ALIGNMENT = void 0;

var _scrollProp, _sizeProp, _positionProp, _marginProp, _oppositeMarginProp;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ALIGNMENT = {
  AUTO: 'auto',
  START: 'start',
  CENTER: 'center',
  END: 'end'
};
exports.ALIGNMENT = ALIGNMENT;
var DIRECTION = {
  HORIZONTAL: 'horizontal',
  VERTICAL: 'vertical'
};
exports.DIRECTION = DIRECTION;
var SCROLL_CHANGE_REASON = {
  OBSERVED: 'observed',
  REQUESTED: 'requested'
};
exports.SCROLL_CHANGE_REASON = SCROLL_CHANGE_REASON;
var scrollProp = (_scrollProp = {}, _defineProperty(_scrollProp, DIRECTION.VERTICAL, 'scrollTop'), _defineProperty(_scrollProp, DIRECTION.HORIZONTAL, 'scrollLeft'), _scrollProp);
exports.scrollProp = scrollProp;
var sizeProp = (_sizeProp = {}, _defineProperty(_sizeProp, DIRECTION.VERTICAL, 'height'), _defineProperty(_sizeProp, DIRECTION.HORIZONTAL, 'width'), _sizeProp);
exports.sizeProp = sizeProp;
var positionProp = (_positionProp = {}, _defineProperty(_positionProp, DIRECTION.VERTICAL, 'top'), _defineProperty(_positionProp, DIRECTION.HORIZONTAL, 'left'), _positionProp);
exports.positionProp = positionProp;
var marginProp = (_marginProp = {}, _defineProperty(_marginProp, DIRECTION.VERTICAL, 'marginTop'), _defineProperty(_marginProp, DIRECTION.HORIZONTAL, 'marginLeft'), _marginProp);
exports.marginProp = marginProp;
var oppositeMarginProp = (_oppositeMarginProp = {}, _defineProperty(_oppositeMarginProp, DIRECTION.VERTICAL, 'marginBottom'), _defineProperty(_oppositeMarginProp, DIRECTION.HORIZONTAL, 'marginRight'), _oppositeMarginProp);
exports.oppositeMarginProp = oppositeMarginProp;
