// template-component-demo/es/Layout/index.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React from 'react';
import Sidebar from './Sidebar';
import styles from './index.module.scss';
export var Layout = function Layout(_ref) {
  var demos = _ref.demos,
      children = _ref.children,
      matchedFilename = _ref.matchedFilename;
  return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
    className: styles.nav
  }, "Docs"), /*#__PURE__*/React.createElement("div", {
    className: styles.container
  }, /*#__PURE__*/React.createElement(Sidebar, {
    items: demos,
    matchedFilename: matchedFilename
  }), /*#__PURE__*/React.createElement("div", {
    className: styles.main
  }, children)));
};
export var BuildLayout = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(BuildLayout, _React$Component);

  function BuildLayout(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _this.handleChangeHash = function () {
      setTimeout(function () {
        // update currentHash after location changed
        _this.setState({
          currentHash: window.location.hash
        });
      }, 0);
    };

    _this.state = {
      currentHash: window.location.hash
    };
    return _this;
  }

  var _proto = BuildLayout.prototype;

  _proto.render = function render() {
    var _this2 = this;

    var _this$props = this.props,
        demos = _this$props.demos,
        children = _this$props.children,
        reactDocs = _this$props.reactDocs;
    var currentHash = this.state.currentHash;
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
      className: styles.layout
    }, children), /*#__PURE__*/React.createElement("ul", {
      className: styles.fixNav
    }, demos.map(function (item) {
      return /*#__PURE__*/React.createElement("li", {
        key: item.filename,
        className: currentHash.replace('#container_', '') === item.filename ? styles.selected : ''
      }, /*#__PURE__*/React.createElement("a", {
        href: "#container_" + item.filename,
        onClick: _this2.handleChangeHash
      }, item.title || item.filename));
    }), reactDocs && reactDocs.length > 0 && /*#__PURE__*/React.createElement("li", null, /*#__PURE__*/React.createElement("a", {
      href: "#container_api",
      onClick: this.handleChangeHash
    }, "API"))));
  };

  return BuildLayout;
}(React.Component);
