// template-component-demo/es/Layout/Sidebar.js
import _inheritsLoose from "@babel/runtime/helpers/inheritsLoose";
import React from 'react';
import { Link } from 'react-router-dom';
import styles from './sidebar.module.scss';

var Sidebar = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Sidebar, _React$Component);

  function Sidebar(props) {
    var _this;

    _this = _React$Component.call(this, props) || this;

    _this.handleChange = function (evt) {
      var value = evt.target.value || '';

      var menus = _this.props.items.filter(function (menu) {
        var item = Array.isArray(menu) ? menu[0] : menu;
        var title = (item.demoKey || item.title).toLocaleLowerCase();
        var val = value.toLocaleLowerCase();
        return title.includes(val);
      });

      _this.setState({
        value: value,
        menus: menus
      });
    };

    _this.state = {
      value: '',
      menus: props.items
    };
    return _this;
  }

  var _proto = Sidebar.prototype;

  _proto.render = function render() {
    var matchedFilename = this.props.matchedFilename;
    var _this$state = this.state,
        value = _this$state.value,
        menus = _this$state.menus;
    return /*#__PURE__*/React.createElement("div", {
      className: styles.sidebar
    }, /*#__PURE__*/React.createElement("div", {
      className: styles.searchbox,
      style: {
        display: 'none'
      }
    }, /*#__PURE__*/React.createElement("input", {
      type: "text",
      className: styles.input,
      placeholder: "Type to search...",
      value: value,
      onChange: this.handleChange
    })), /*#__PURE__*/React.createElement("dl", {
      className: styles.menuList
    }, menus.map(function (menu, key) {
      var item = Array.isArray(menu) ? menu[0] : menu;
      return /*#__PURE__*/React.createElement("dt", {
        className: "" + styles.item + (matchedFilename === (item.demoKey || item.filename) ? " " + styles.selected : ''),
        key: (item.demoKey || item.filename) + "-" + key
      }, /*#__PURE__*/React.createElement(Link, {
        to: "?demo=" + (item.demoKey || item.filename)
      }, item.demoKey || item.title));
    })));
  };

  return Sidebar;
}(React.Component);

export { Sidebar as default };
